Add datetime: null when an item has start_datetime and end_datetime but no datetime.

krisaoe opened the pull request:
Ensures STAC spec compliance by adding datetime: null when temporal ranges are present. See #158

krisaoe committed:
Add datetime: null when an item has start_datetime and end_datetime but no datetime.

gadomski reviewed (changes requested):
It'd be good to add a test to confirm the expected behavior. I'm also summoning @bitner in case he's got insight into the fix strategy.

krisaoe committed:
Move datetime:null fix to jsonb content_hydrate to cover all hydration paths

krisaoe committed:
Add pgtap and basic tests for datetime:null hydration

krisaoe committed:
Update CHANGELOG.md with fix and linked issue

krisaoe commented:
Moved the fix to the jsonb overload - all hydration paths flow through it (including tests).
Added 2 pgtap tests (adds datetime key when start/end_datetime present & preserves existing datetime), added a hydration basic test to test the full flow.

gadomski committed:
Merge branch 'main' into fix/allow-datetime-null-if-range-exists

bitner commented:
@krisaoe the only things that we strip out of an items json on ingest are the id, geometry, collection, and type. See the content_slim function in 003a_items.sql. We leave (correctly or incorrectly) datetime, start_datetime, and end_datetime exactly as they were in the original json. In very early versions, we stripped out those fields and reconstructed them from the instantiated datetime and end_datetime fields in the database on output, but there were instances where when we re-put those back into the output json that there were issues with null handling (as you are having) or with precision, so we made the decision then that we would keep those fields within the json (at the expense of duplicating that storage) to make sure that things were faithfully round tripped.

PGStac does not try to do any validating of stac records and expects that any validation is done as part of an ingest workflow.

Can you verify if the behavior I am describing above is what you are seeing or if you are seeing something come out of pgstac with json that does not match what had been ingested into pgstac?

vincentsarago commented:
FYI there was a bug in stac-pydantic/stac-fastapi which was stripping null datetime, but this is now resolved https://github.com/stac-utils/stac-pydantic/pull/190

krisaoe committed:
Revert content_hydrate datetime:null patch in favour of root-cause fix

krisaoe committed:
Preserve JSON null values in merge_jsonb and strip_jsonb
